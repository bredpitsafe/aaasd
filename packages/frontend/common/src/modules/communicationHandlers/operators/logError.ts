import { tapError } from '@common/rx';

import type { TSendEnvelope } from '../../../lib/BFFSocket/def';
import type { SocketStreamError } from '../../../lib/SocketStream/SocketStreamError';
import { logger } from '../../../utils/Tracing';

export function logErrorOperator(request: TSendEnvelope<any>) {
    return tapError((err: SocketStreamError) => {
        logger.error(`[FetchHandlers]: ${err.message}`, {
            socketURL: err.socketURL,
            reason: err.reason,
            traceId: err.traceId,
            timestamp: err.timestamp,
            correlationId: err.correlationId,
            requestType: request.payload.type,
            code: err.code,
            args: err.args,
        });
    });
}
