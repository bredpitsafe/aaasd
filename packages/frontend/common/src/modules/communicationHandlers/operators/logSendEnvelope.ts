import type { TSendEnvelope } from '../../../lib/BFFSocket/def';
import type { TSocketURL } from '../../../types/domain/sockets';
import { logger } from '../../../utils/Tracing';
import { prepareEnvelopeToTracing } from '../utils';

export function logSendEnvelope<T extends TSendEnvelope<any>>(url: TSocketURL, envelope: T) {
    logger.trace('[FetchHandlers]: Send envelope', {
        url,
        envelope: prepareEnvelopeToTracing(envelope),
        type: envelope.payload.type,
    });
}
