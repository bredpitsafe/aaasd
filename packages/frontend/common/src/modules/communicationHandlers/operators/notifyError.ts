import { tapError } from '@common/rx';
import type { Milliseconds } from '@common/types';

import { SocketStreamError } from '../../../lib/SocketStream/SocketStreamError';
import type { IModuleNotifications } from '../../notifications/def';

export function notifyErrorOperator(notification: IModuleNotifications) {
    return tapError((err) => {
        if (err instanceof SocketStreamError) {
            return notification.error({
                traceId: err.traceId,
                message: err.message,
                description: err.description,
                timestamp: err.timestamp as Milliseconds,
                socketURL: err.socketURL,
            });
        }
        return notification.error({
            message: err.message,
            description: err.stack,
        });
    });
}
