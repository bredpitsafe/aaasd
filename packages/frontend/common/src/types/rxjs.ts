import type { Observable, Subject, Subscription } from 'rxjs';

export interface IObserver<T, E> {
    next: (value: T) => void;
    error: (err: E) => void;
    complete: () => void;
}

export interface IObservable<T, E> extends Observable<T> {
    subscribe(next: (value: T) => void): Subscription;
    subscribe(observer?: Partial<IObserver<T, E>>): Subscription;
}

export interface ISubject<T, E> extends Subject<T> {
    error(err: E): void;
    subscribe(next: (value: T) => void): Subscription;
    subscribe(observer?: Partial<IObserver<T, E>>): Subscription;
}
