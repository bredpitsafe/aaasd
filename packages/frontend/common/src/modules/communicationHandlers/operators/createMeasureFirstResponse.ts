import type { Milliseconds } from '@common/types';
import { iso2milliseconds, sum } from '@common/utils';
import { once } from 'lodash-es';
import { tap } from 'rxjs/operators';

import type { TReceivedData } from '../../../lib/BFFSocket/def';
import type { TSocketURL } from '../../../types/domain/sockets';
import type {
    ExtractSyncedValueDescriptor,
    TValueDescriptor2,
} from '../../../utils/ValueDescriptor/types';
import { isSyncedValueDescriptor } from '../../../utils/ValueDescriptor/utils';

export function createMeasureFirstResponse(
    url: TSocketURL,
    collect: (v: { labels: [TSocketURL]; observe: Milliseconds }) => void,
) {
    let sendTime = <Milliseconds>0;

    const initFirstResponse = (timestamp: Milliseconds) => {
        sendTime = timestamp;
    };

    const measureFirstResponseOperator = <
        T extends TValueDescriptor2<TReceivedData<unknown>>,
    >() => {
        const collectOnce = once((vd: ExtractSyncedValueDescriptor<T>) =>
            collect({
                labels: [url],
                observe: sum(iso2milliseconds(vd.value.timestamp), -sendTime),
            }),
        );

        return tap<T>((envelope) => {
            if (isSyncedValueDescriptor(envelope)) {
                collectOnce(envelope);
            }
        });
    };

    return { initFirstResponse, measureFirstResponseOperator };
}
