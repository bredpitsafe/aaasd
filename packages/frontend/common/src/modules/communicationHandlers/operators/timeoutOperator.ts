import type { Observable } from 'rxjs';
import { timeout } from 'rxjs';

import type { TSendEnvelope } from '../../../lib/BFFSocket/def';
import { EErrorReason } from '../../../lib/SocketStream/def';
import { SocketStreamError } from '../../../lib/SocketStream/SocketStreamError';
import type { TSocketURL } from '../../../types/domain/sockets.ts';
import { EGrpcErrorCode } from '../../../types/GrpcError';
import type { TRequiredHandlerOptions } from '../def';

export function timeoutOperator<T>(
    url: TSocketURL,
    envelope: TSendEnvelope<any>,
    options: TRequiredHandlerOptions,
) {
    return timeout<T, Observable<T>>({
        first: options.timeout,
        with: () => {
            throw new SocketStreamError('Server message receive timeout', {
                code: EGrpcErrorCode.UNKNOWN,
                reason: EErrorReason.timeout,
                traceId: envelope.traceId,
                correlationId: envelope.correlationId,
                socketURL: url,
            });
        },
    });
}
