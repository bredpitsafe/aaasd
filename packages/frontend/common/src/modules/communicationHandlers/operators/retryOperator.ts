import type { Observable } from 'rxjs';
import { concat, mergeMap, of, throwError, timer } from 'rxjs';
import { catchError } from 'rxjs/operators';

import { EErrorReason } from '../../../lib/SocketStream/def';
import { SocketStreamError } from '../../../lib/SocketStream/SocketStreamError';
import { convertErrToGrpcFail } from '../../../utils/ValueDescriptor/Fails';
import type { TValueDescriptor2 } from '../../../utils/ValueDescriptor/types';
import { createUnsyncedValueDescriptor } from '../../../utils/ValueDescriptor/utils';
import type { TRequiredHandlerOptions } from '../def';

const UNSPECIFIED_RETRY_DELAY = 5_000;

export function retryOperator<T extends TValueDescriptor2<any>>(options: TRequiredHandlerOptions) {
    let _index = 0;
    return catchError<T, Observable<T>>((error: Error | SocketStreamError, caught) => {
        const index = _index++;
        const isCommunicatorError = error instanceof SocketStreamError;

        if (!isCommunicatorError) {
            return throwError(error);
        }

        const isRetryOnReconnect =
            options.retryOnReconnect && error.reason === EErrorReason.socketClose;
        const isRetryOnTimeout = index < options.retries && error.reason === EErrorReason.timeout;
        // retry unspecified error only once
        const isRetryOnUnspecified =
            isCommunicatorError && index === 0 && error.kind === 'Unspecified';

        if (isRetryOnUnspecified) {
            return concat(
                of(createUnsyncedValueDescriptor(convertErrToGrpcFail(error)) as T),
                timer(UNSPECIFIED_RETRY_DELAY).pipe(mergeMap(() => caught)),
            );
        }

        if (isRetryOnReconnect || isRetryOnTimeout) {
            return concat(
                of(createUnsyncedValueDescriptor(convertErrToGrpcFail(error)) as T),
                caught,
            );
        }

        return throwError(error);
    });
}
