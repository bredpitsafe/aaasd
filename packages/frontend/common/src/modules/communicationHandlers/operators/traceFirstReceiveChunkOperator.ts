import { once } from 'lodash-es';
import { tap } from 'rxjs/operators';

import type { TReceivedData } from '../../../lib/BFFSocket/def';
import type { TSocketURL } from '../../../types/domain/sockets';
import { logger } from '../../../utils/Tracing';
import type { TValueDescriptor2 } from '../../../utils/ValueDescriptor/types';
import { isReceivingValueDescriptor } from '../../../utils/ValueDescriptor/utils';

export function traceFirstReceiveChunkOperator<T extends TValueDescriptor2<TReceivedData<unknown>>>(
    url: TSocketURL,
    type: string,
) {
    const logOnce = once(() =>
        logger.trace('[FetchHandlers]: Receive first chunk', {
            url,
            requestType: type,
        }),
    );

    return tap<T>((vd) => {
        if (isReceivingValueDescriptor(vd)) {
            logOnce();
        }
    });
}
