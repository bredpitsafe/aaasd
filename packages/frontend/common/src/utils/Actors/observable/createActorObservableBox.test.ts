import { cloneDeep } from 'lodash-es';
import { firstValueFrom, from, interval, lastValueFrom, Observable, of } from 'rxjs';
import { finalize, map, take, tap } from 'rxjs/operators';

import type { TActorObservableBox } from './def.ts';
import { createActorObservableBox } from './index.ts';

const BOX_TYPE = 'TEST_BOX';

const actor = {};
type Request = { data: any };
type Response = { data: any };

describe('createActorObservableBox', () => {
    let box: TActorObservableBox<'TEST_BOX', Request, Response>;
    const request: Request = { data: {} };
    const response: Response = { data: {} };

    describe('subscriptions behavior', () => {
        beforeEach(() => {
            box = createActorObservableBox<Request, Response>()(BOX_TYPE);
        });

        afterEach(() => {
            box.destroy();
        });

        it('response stream should receive data from request stream', (done) => {
            box.responseStream(actor, (req) => {
                expect(req).toBe(request);
                done();
                return of(response);
            });

            box.requestStream(actor, request).subscribe();
        });

        it('response stream should return value to request stream', (done) => {
            box.responseStream(actor, () => {
                return of(response);
            });

            box.requestStream(actor, request).subscribe((value) => {
                expect(value).toBe(response);
                done();
            });
        });

        it('response stream should handle multiple requests', async () => {
            const cb = jest.fn();
            box.responseStream(actor, (req) => {
                cb(req);
                return of(response);
            });

            await firstValueFrom(box.requestStream(actor, request));
            await firstValueFrom(box.requestStream(actor, request));
            expect(cb).toHaveBeenCalledTimes(2);
        });

        it('response stream should handle multiple responses', async () => {
            const responses = [response, response, response];
            box.responseStream(actor, () => {
                return from(responses);
            });

            const cb = jest.fn();
            await lastValueFrom(box.requestStream(actor, request).pipe(tap(cb)));
            expect(cb).toHaveBeenCalledTimes(responses.length);
            responses.forEach((response, i) => {
                expect(response).toBe(cb.mock.calls[i][0]);
            });
        });

        it('errors requestStream when responseStream observable ends with error', async () => {
            const err = new Error('test error');
            box.responseStream(actor, () => {
                return new Observable((subscriber) => {
                    subscriber.next(response);
                    subscriber.error(err);
                });
            });

            await expect(lastValueFrom(box.requestStream(actor, request))).rejects.toBe(err);
        });

        it('errors requestStream when responseStream is not registered', async () => {
            await expect(
                lastValueFrom(box.requestStream(actor, request)),
            ).rejects.toMatchSnapshot();
        });

        it('errors when trying to reevaluate responseStream', () => {
            box.responseStream(actor, () => of(response));
            expect(() =>
                box.responseStream(actor, () => of(response)),
            ).toThrowErrorMatchingSnapshot();
            box.destroy();
        });

        it('deregisters responseStream', async () => {
            const deregisterResponseStream = box.responseStream(actor, () => of(response));
            deregisterResponseStream();
            await expect(
                firstValueFrom(box.requestStream(actor, request)),
            ).rejects.toMatchSnapshot();
            expect(() => box.responseStream(actor, () => of(response))).not.toThrow();
            await expect(firstValueFrom(box.requestStream(actor, request))).resolves.toBe(response);
        });

        it('handles multiple requestStreams with a single responseStream', async () => {
            const responses: Response[] = [
                cloneDeep(response),
                cloneDeep(response),
                cloneDeep(response),
            ];
            let responseIndex = 0;
            box.responseStream(actor, () => {
                return of(responses[responseIndex++]);
            });

            for (const res of responses) {
                const streamRes = await lastValueFrom(box.requestStream(actor, request));
                expect(streamRes).toBe(res);
            }
        });

        it('completes responseStream when requestStream is completed', async () => {
            const cbComplete = jest.fn();
            const obs$ = interval(1).pipe(
                map(() => response),
                finalize(cbComplete),
            );
            box.responseStream(actor, () => obs$);

            await lastValueFrom(box.requestStream(actor, request).pipe(take(10)));

            expect(cbComplete).toHaveBeenCalled();
        });

        it('completes responseStream when requestStream ends with error', async () => {
            const err = new Error('test error');
            const cbComplete = jest.fn();
            const obs$ = interval(1).pipe(
                map(() => response),
                finalize(cbComplete),
            );
            box.responseStream(actor, () => obs$);

            await expect(
                lastValueFrom(
                    box.requestStream(actor, request).pipe(
                        map((res, i) => {
                            if (i === 10) {
                                throw err;
                            }
                            return res;
                        }),
                    ),
                ),
            ).rejects.toBe(err);

            expect(cbComplete).toHaveBeenCalled();
        });

        it('completes only required response observable when requestStream completes', async () => {
            const cbComplete1 = jest.fn();
            const obs$1 = interval(1).pipe(
                map(() => response),
                finalize(cbComplete1),
            );
            const cbComplete2 = jest.fn();
            const obs$2 = interval(1).pipe(
                map(() => response),
                finalize(cbComplete2),
            );
            let requestIndex = 0;
            box.responseStream(actor, () => {
                if (requestIndex++ === 0) {
                    return obs$1;
                }
                return obs$2;
            });

            box.requestStream(actor, request).subscribe();
            await lastValueFrom(box.requestStream(actor, request).pipe(take(10)));

            expect(cbComplete1).not.toHaveBeenCalled();
            expect(cbComplete2).toHaveBeenCalled();
        });

        it('errors only a single requestStream when multiple streams are active', async () => {
            const err = new Error('test error');

            let responseIndex = 0;
            box.responseStream(actor, () => {
                return new Observable((subscriber) => {
                    subscriber.next(response);

                    if (responseIndex === 0) {
                        subscriber.error(err);
                    }
                    responseIndex++;
                });
            });

            await expect(lastValueFrom(box.requestStream(actor, request))).rejects.toBe(err);

            const cbNext = jest.fn();
            const cbComplete = jest.fn();
            const subscription = box
                .requestStream(actor, request)
                .pipe(tap(cbNext), finalize(cbComplete))
                .subscribe();

            expect(cbNext).toHaveBeenLastCalledWith(response);
            expect(cbComplete).not.toHaveBeenCalled();
            subscription.unsubscribe();
        });

        it('completes only a single requestStream when multiple streams are active', async () => {
            let responseIndex = 0;
            box.responseStream(actor, () => {
                return new Observable((subscriber) => {
                    subscriber.next(response);

                    if (responseIndex === 0) {
                        subscriber.complete();
                    }
                    responseIndex++;
                });
            });
            expect(await lastValueFrom(box.requestStream(actor, request))).toBe(response);

            const cbNext = jest.fn();
            const cbComplete = jest.fn();
            const subscription = box
                .requestStream(actor, request)
                .pipe(tap(cbNext), finalize(cbComplete))
                .subscribe();

            expect(cbNext).toHaveBeenLastCalledWith(response);
            expect(cbComplete).not.toHaveBeenCalled();
            subscription.unsubscribe();
        });

        it('requestStream subscriptions work after emitting an error before', async () => {
            const err = new Error('test error');

            let responseIndex = 0;
            box.responseStream(actor, () => {
                return new Observable((subscriber) => {
                    if (responseIndex === 0) {
                        subscriber.error(err);
                        responseIndex++;
                        return;
                    }
                    subscriber.next(response);
                });
            });

            await expect(firstValueFrom(box.requestStream(actor, request))).rejects.toBe(err);
            await expect(firstValueFrom(box.requestStream(actor, request))).resolves.toBe(response);
        });
    });

    describe('box deduplication behavior', () => {
        beforeEach(() => {
            box = createActorObservableBox<Request, Response>()(BOX_TYPE);
        });

        it('creating a box multiple times should yield the same box', () => {
            const box2 = createActorObservableBox<Request, Response>()(BOX_TYPE);
            expect(box2).toBe(box);
            box.destroy();
        });

        it('trying to subscribe to a deduplicated box yields subscription results', async () => {
            box.responseStream(actor, () => of(response));
            const box2 = createActorObservableBox<Request, Response>()(BOX_TYPE);
            expect(await lastValueFrom(box2.requestStream(actor, request))).toBe(response);
            box.destroy();
        });

        it('should destroy a box', () => {
            box.destroy();
            const box2 = createActorObservableBox<Request, Response>()(BOX_TYPE);
            expect(box2).not.toBe(box);
        });

        it('should destroy all deduplicated boxes', () => {
            const box2 = createActorObservableBox<Request, Response>()(BOX_TYPE);
            box.destroy();

            const box3 = createActorObservableBox<Request, Response>()(BOX_TYPE);
            expect(box3).not.toBe(box2);
        });

        it('should finalize all box subscriptions upon box destruction', async () => {
            const cbFinalize = jest.fn();
            box.responseStream(actor, () =>
                new Observable<Response>((subscriber) => subscriber.next(response)).pipe(
                    finalize(cbFinalize),
                ),
            );

            const promise = lastValueFrom(box.requestStream(actor, request));
            box.destroy();
            expect(await promise).toBe(response);
            expect(cbFinalize).toHaveBeenCalled();
        });
    });
});
