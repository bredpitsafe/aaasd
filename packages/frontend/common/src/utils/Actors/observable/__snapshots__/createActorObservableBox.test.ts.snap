// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`createActorObservableBox subscriptions behavior deregisters responseStream 1`] = `[Error: responseStream is not yet registered]`;

exports[`createActorObservableBox subscriptions behavior errors requestStream when responseStream is not registered 1`] = `[Error: responseStream is not yet registered]`;

exports[`createActorObservableBox subscriptions behavior errors when trying to reevaluate responseStream 1`] = `"responseStream is already registered for box type 'TEST_BOX'"`;
